src/http/http.ml

```diff
@@ -837,6 +837,16 @@ let run
       ignore
   in
 
+  let prev_signal_behavior = ref Sys.Signal_default in
```

Comment:
Try combining this with the next line with

```ocaml
let rec prev_signal_behavior := Sys.signal ...
```

Not sure if it will work.

Reply:
Nope, that's a syntax error

Reply:
Ah yes, I meant

```ocaml
let rec prev_signal_behavior = ref (Sys.signal ...
```

Reply:
Ah, but is that possible with the new structure?
```ocaml
let prev_signal_behavior = ref Sys.Signal_default in

let handler () = (Sys.Signal_handle (fun signal ->
  restore_terminal ();
  match !prev_signal_behavior with
    | Sys.Signal_default -> exit 1
    | Sys.Signal_handle f -> f signal
    | Sys.Signal_ignore -> ignore ()
)) in

prev_signal_behavior := Sys.signal Sys.sigint @@ handler ();
prev_signal_behavior := Sys.signal Sys.sigterm @@ handler ();
```

Reply:
This new structure is incorrect AFAIK, as the derferencing of `prev_signal_behavior` is suspended inside closures (`fun signal -> ...`), which are only called upon actual reception of signals. By then, the setup code will have assigned `prev_signal_behavior` twice, the second time for SIGTERM, completely losing the SIGINT previous behavior, and making both signals behave as the previous behavior of SIGTERM.

By far the easiest way to get this code right (and to make it immediately legible for readers) is to use a local reference for each signal, and you will probably be able to use the `rec` in that case, too.

Reply:
I'm not exactly sure what you mean by using "a local reference for each signal". It seems to me that the ref has to be passed as an argument to some sort of `create_handler` function, no?
```ocaml
let create_handler prev_signal_behavior = (Sys.Signal_handle (fun signal ->
  restore_terminal ();
  match !prev_signal_behavior with
    | Sys.Signal_default -> exit 1
    | Sys.Signal_handle f -> f signal
    | Sys.Signal_ignore -> ignore ()
)) in

let rec sigint_behavior = ref (Sys.signal Sys.sigint @@ handler sigint_behavior);
```
Except it looks like this rec is also not possible, I'm getting a compilation error:
```
This kind of expression is not allowed as right-hand side of `let rec'ocamllsp
```

Reply:
No, the `ref` can and really should be created inside the `create_handler` function, which guarantees that it will be local to that handler. `create_handler` should take a signal number, not a ref to a signal behavior.